### PR TITLE
add custom relu implementation

### PR DESCRIPTION
**Pull request by be-apt**
thanks to @matbesancon for #91, added a very trivial example of implementing ReLU as a layer in a NN trained on MNIST dataset using Flux.jl
There are some issues pending, specifically training on the complete dataset if very slow, and gives a JuMP warning:
```
Warning: The addition operator has been used on JuMP expressions a large number of times. 
This warning is safe to ignore but may indicate that model generation is slower than necessary. 
For performance reasons, you should not add expressions in a loop.
Instead of x += , use append!(x,y) to modify x in place. 
If y is a single variable, you may also use push!(x, coef, y) in place of x += coef*y
```

### CODE REVIEW

**Comment by be-apt** [comment]
also, current code works on latest Julia build (1.6.0). Seems Julia 1.0 supports only Tracker based Flux (while the latest one supports Zygote.jl)

**Comment by matbesancon** [comment]
we should make this into a documentation page instead I think, it will also build it at every PR

File: examples/custom-relu-mnist.jl [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+            ) 
+
+            DiffOpt.backward(model)  # find grad
+
+            dy[:, i] = MOI.get.(
+                model,
+                DiffOpt.BackwardOut{DiffOpt.LinearObjective}(), 
+                x, 
+            )  # coeff of `x` in -2x'y
+            dy[:, i] = -2 * dy[:, i]
```
Comment (odow) [inline examples/custom-relu-mnist.jl:75]:
```suggestion
            dy[:, i] .= -2 .* dy[:, i]
```

File: examples/custom-relu-mnist.jl [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+    relu method for a Matrix
+"""
+function myRelu(y::AbstractMatrix{T}; model = Model(() -> diff_optimizer(OSQP.Optimizer))) where {T}
+    x̂ = zero(y)
+    
+    # model init
+    N = length(y[:, 1])
+    empty!(model)
+    set_optimizer_attribute(model, MOI.Silent(), true)
+    @variable(model, x[1:N] >= zero(T))
```
Comment (odow) [inline examples/custom-relu-mnist.jl:38]:
```suggestion
    @variable(model, x[1:N] >= 0)
```

File: examples/custom-relu-mnist.jl [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+"""
+    relu method for a Matrix
+"""
+function myRelu(y::AbstractMatrix{T}; model = Model(() -> diff_optimizer(OSQP.Optimizer))) where {T}
+    x̂ = zero(y)
+    
+    # model init
+    N = length(y[:, 1])
+    empty!(model)
+    set_optimizer_attribute(model, MOI.Silent(), true)
```
Comment (odow) [inline examples/custom-relu-mnist.jl:37]:
```suggestion
    set_silent(model)
```

File: examples/custom-relu-mnist.jl [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+    @variable(model, x[1:N] >= zero(T))
+    
+    for i in 1:size(y)[2]
```
Comment (odow) [inline examples/custom-relu-mnist.jl:40]:
```suggestion
    for i in 1:size(y, 2)
```

**Review by odow — COMMENTED** [review]
The warning suggests you are doing some arithmetic outside the JuMP macros. I couldn't see anything obvious here, so it must be somewhere inside DiffOpt.jl.

File: examples/custom-relu-mnist.jl [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+    N = length(y[:, 1])
+    empty!(model)
+    set_silent(model)
+    @variable(model, x[1:N] >= 0)
+    
+    for i in 1:size(y, 2)
+        @objective(
+            model,
+            Min,
+            x'x -2x'y[:, i]
```
Comment (blegat) [inline examples/custom-relu-mnist.jl:44]: This is where the warning comes from I believe, this should be fixed by https://github.com/jump-dev/MutableArithmetics.jl/pull/87

**Comment by blegat** [comment]
> we should make this into a documentation page instead I think, it will also build it at every PR

I agree, it would be helpful to run it at every commit with Literate

File: examples/custom-relu-mnist.jl (hunk 5 above)
Comment (be-apt) [reply examples/custom-relu-mnist.jl]: solved. seems there's a `set_objective_coefficient` method. + speeds up training a bit

File: examples/custom-relu-mnist.jl [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+
+# float64 to float16, to save memory
+X = convert(Array{Float16,2}, X) 
+test_X = convert(Array{Float16,2}, test_X)
+
+X = X[:, 1:10000]
+Y = Y[:, 1:10000];
+
+"""
+    relu method for a Matrix
```
Comment (matbesancon) [inline examples/custom-relu-mnist.jl:29]:
```suggestion
relu method for a Matrix
```

File: examples/custom-relu-mnist.jl [hunk 7]
```diff
@@ -0,0 +1,103 @@
...
+"""
+function myRelu(y::AbstractMatrix{T}; model = Model(() -> diff_optimizer(OSQP.Optimizer))) where {T}
```
Comment (matbesancon) [inline examples/custom-relu-mnist.jl:31]: `matrix_relu` better than `myRelu`? Also write out the model in the docstring

File: examples/custom-relu-mnist.jl [hunk 8]
```diff
@@ -0,0 +1,103 @@
...
+        return (NO_FIELDS, dy)
+    end
+    return pv, pullback_myRelu
+end
+
+m = Chain(
+    Dense(784, 64),
+    myRelu,
+    Dense(64, 10),
+    softmax
```
Comment (matbesancon) [inline examples/custom-relu-mnist.jl:88]:
```suggestion
    softmax,
```

File: examples/custom-relu-mnist.jl (hunk 7 above)
Comment (be-apt) [reply examples/custom-relu-mnist.jl:31]: changing it

**Comment by matbesancon** [comment]
status on this one? Should we merge it or make it a doc page first?

**Comment by be-apt** [comment]
@matbesancon should we close this (without merging ?)
https://github.com/jump-dev/DiffOpt.jl/pull/95 already contains the changes introduced in this PR

**Comment by matbesancon** [comment]
#95 will be lighter to review if this gets merged first.
Things missing:
- Some comments / explanations
- Adding it to the tests

**Comment by matbesancon** [comment]
otherwise there is a risk we have this example breaking at some point

**Comment by matbesancon** [comment]
bump @be-apt to add the example to CI

**Comment by be-apt** [comment]
closing this. duplicate code of #95

**Pull request closed without merging**